### PR TITLE
[FIX] pos_loyalty: Check loyalty product availability before opening

### PR DESCRIPTION
**Pull request by lse-odoo**
Before this commit:
Traceback in PoS session if a loyalty product is archived.

After this commit:
Prevent PoS opening by announcing wrongly configured products

OPW-2276745

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/point_of_sale/models/pos_config.py [hunk 1]
```diff
@@ -355,6 +355,23 @@ def _check_payment_method_receivable_accounts(self):
...
+        if self.module_pos_loyalty and self.loyalty_id:
+            archived_reward_products = {}
+            for loyalty in self.loyalty_id:
+                for reward in loyalty.reward_ids.filtered(
+                        lambda r: r.reward_type == "resale" and r.point_product_id and not r.point_product_id.active):
+                    archived_reward_products.setdefault(loyalty, []).append(reward.point_product_id)
+            if archived_reward_products:
+                t = ""
+                for loyalty, products in archived_reward_products.items():
+                    t += "\nLoyalty \"%s\" (id %s):" % (loyalty.name, loyalty.id)
```
Comment (fmdl) [inline addons/point_of_sale/models/pos_config.py:369]: `t += _("\nLoyalty \"%s\" (id %s):") % (loyalty.name, loyalty.id)`
use _() for translation

Comment (lse-odoo) [reply addons/point_of_sale/models/pos_config.py]: done with the updated version. Thank you for the proposition

**Comment by lse-odoo** [comment]
@nim-odoo Could you review it please ?

File: addons/point_of_sale/models/pos_config.py [hunk 2]
```diff
@@ -355,6 +355,23 @@ def _check_payment_method_receivable_accounts(self):
...
+            for loyalty in self.loyalty_id:
+                for reward in loyalty.reward_ids.filtered(
+                        lambda r: r.reward_type == "resale" and r.point_product_id and not r.point_product_id.active):
+                    archived_reward_products.setdefault(loyalty, []).append(reward.point_product_id)
+            if archived_reward_products:
+                t = ""
+                for loyalty, products in archived_reward_products.items():
+                    t += "\n" + _("Loyalty") + ' "%s" (id %s):' % (loyalty.name, loyalty.id)
+                    for product in products:
+                        t += "\n - " + _("Product") + ' "%s" (id %s)' % (product.name, product.id)
```
Comment (fmdl) [inline addons/point_of_sale/models/pos_config.py:371]: use `display_name` it is better

File: addons/point_of_sale/models/pos_config.py [hunk 3]
```diff
@@ -355,6 +355,23 @@ def _check_payment_method_receivable_accounts(self):
                 _("You must configure an intermediary account for the payment methods: %s.") % method_names
             )
 
+    @api.constrains('module_pos_loyalty')
+    def _check_loyalties_product_availability(self):
+        if self.module_pos_loyalty and self.loyalty_id:
```
Comment (fmdl) [inline addons/point_of_sale/models/pos_config.py:360]: more simpler and work with multi records
```python
def _check_loyalties_product_availability(self):
    archived_products = self.mapped('loyalty_id.reward_ids.point_product_id').filtered(
                        lambda p: not p.active).
   if archived_products:
       raise ValidationError(
                    _('The Discount Product of the following Loyalty Program are not available in the PoS.\n') + '\n'.join(archived_products.mapped('display_name') )
```

Comment (lse-odoo) [reply addons/point_of_sale/models/pos_config.py]: Updated.
But I still prefer to show the id of the products as several products may have the same name (which was the case on the client database).
I also modify the message as the client may want to change to another product his loyalty.
Thank you for your proposition

**Review by nim-odoo — CHANGES_REQUESTED** [review]
`loyalty_id` is defined in module `pos_loyalty`.

File: addons/point_of_sale/models/pos_config.py [hunk 4]
```diff
@@ -355,6 +355,17 @@ def _check_payment_method_receivable_accounts(self):
                 _("You must configure an intermediary account for the payment methods: %s.") % method_names
             )
 
+    @api.constrains('module_pos_loyalty')
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_config.py:358]: That's not really what I had in mind. you should move the check in the `pos_loyalty` module. Btw, the commit name should be: `[FIX] pos_loyalty: ...`

**Comment by lse-odoo** [comment]
@nim-odoo updated with PR in enterprise code: https://github.com/odoo/enterprise/pull/11255

File: addons/point_of_sale/models/pos_config.py [hunk 5]
```diff
@@ -555,6 +561,8 @@ def open_session_cb(self):
             self._check_company_payment()
             self._check_currencies()
             self._check_payment_method_receivable_accounts()
+            self._check_loyalties_product_availability()
```
Comment (nim-odoo) [inline addons/point_of_sale/models/pos_config.py:564]: This should be moved to `pos_loyalty` module somehow. Your check should be done in `pos_loyalty`; `point_of_sale` doesn't need to be aware of them.

For example, in `pos_loyalty`:
```
    def open_session_cb(self):
        self.ensure_one()
        if not self.current_session_id:
            self._check_loyalties_product_availability()
        return super().open_session_cb()
```

Ideally, there should not be any change in `point_of_sale`

Comment (lse-odoo) [reply addons/point_of_sale/models/pos_config.py:564]: Got it, but the "somehow" part is harder than I thought as no models in `pos_loyalty` inherit from the `pos_config`. Do you have any idea on the matter or should I forward it to your team and look at it at your own pace ?

Comment (nim-odoo) [reply addons/point_of_sale/models/pos_config.py:564]: You can create a new class which inherits from the right model. For example like this: https://github.com/odoo/enterprise/blob/13.0/sale_intrastat/models/account_move.py

**Pull request closed without merging**